File: Packages/TeamTasksKit/Sources/TeamTasksSupabase/Services/SupabaseProfileService.swift
import Foundation
import TeamTasksCore

/// `ProfileService` on PostgREST (docs/CONTRACTS.md §4.3; docs/CONTRACTS-V2.md §2, §5, §9).
struct SupabaseProfileService: ProfileService {
    let context: SupabaseContext

    /// No visible profile for the session user means the account no longer exists → `.notAuthenticated`.
    /// v2: with the avatar and the onboarding fields (`onboardedAt`, `createdAt`), which only this read fills.
    func myProfile() async throws -> UserProfile {
        let rows = try await context.rest.fetch([ProfileRow].self) { RestQuery.myProfile(me: $0.userId) }
        guard let row = rows.first else { throw AppError.notAuthenticated }
        return row.profile
    }

    /// `PATCH profiles?id=eq.<me>` with the trimmed name; 0 rows updated → `.forbidden`. v2: the result keeps the
    /// avatar.
    func updateDisplayName(_ name: String) async throws -> UserProfile {
        let name = try InputValidation.displayName(name)
        let rows = try await context.rest.fetch([ProfileRow].self) {
            RestQuery.updateDisplayName(me: $0.userId, name: name)
        }
        guard let row = rows.first else { throw AppError.forbidden }
        return row.profile
    }

    /// v2: `PATCH profiles?id=eq.<me>` of both avatar columns (docs/CONTRACTS-V2.md §2). The emoji is checked and
    /// normalized first (nil or blank = the initials; a typed color is always valid); 0 rows updated → `.forbidden`.
    func updateAvatar(color: ColorKey?, emoji: String?) async throws -> UserProfile {
        let emoji = try InputValidation.emoji(emoji)
        let rows = try await context.rest.fetch([ProfileRow].self) {
            RestQuery.updateAvatar(me: $0.userId, color: color, emoji: emoji)
        }
        guard let row = rows.first else { throw AppError.forbidden }
        return row.profile
    }

    /// v2: `complete_onboarding()` sets `onboarded_at` once; later calls write nothing.
    func completeOnboarding() async throws {
        _ = try await context.rest.send { _ in RestQuery.rpc("complete_onboarding") }
    }
}

/// `PushService` on the `enable_push` / `disable_push` RPCs and `push_subscriptions` (docs/CONTRACTS.md §4, §7).
struct SupabasePushService: PushService {
    let context: SupabaseContext

    func currentTopic() async throws -> String? {
        let rows = try await context.rest.fetch([PushTopicRow].self) { RestQuery.pushTopic(me: $0.userId) }
        return rows.first?.topic
    }

    func enable() async throws -> String {
        try await context.rest.fetch(String.self) { _ in RestQuery.rpc("enable_push") }
    }

    func disable() async throws {
        _ = try await context.rest.send { _ in RestQuery.rpc("disable_push") }
    }
}

File: android/supabase/src/main/kotlin/io/github/notkanaa/equipe/supabase/SupabaseProfileService.kt
package io.github.notkanaa.equipe.supabase

import io.github.notkanaa.equipe.core.AppError
import io.github.notkanaa.equipe.core.InputValidation
import io.github.notkanaa.equipe.core.ProfileService
import io.github.notkanaa.equipe.core.PushService
import io.github.notkanaa.equipe.core.UserProfile

/** [ProfileService] on PostgREST (docs/CONTRACTS.md §4.3). Port of SupabaseProfileService.swift. */
internal class SupabaseProfileService(private val context: SupabaseContext) : ProfileService {
    /** No visible profile for the session user means the account no longer exists → [AppError.NotAuthenticated]. */
    override suspend fun myProfile(): UserProfile {
        val rows = context.rest.fetchAll(ProfileRow::decode) { RestQuery.myProfile(it.userId) }
        return rows.firstOrNull()?.profile ?: throw AppError.NotAuthenticated
    }

    /** `PATCH profiles?id=eq.<me>` with the trimmed name; 0 rows updated → [AppError.Forbidden]. */
    override suspend fun updateDisplayName(name: String): UserProfile {
        val validName = InputValidation.displayName(name)
        val rows = context.rest.fetchAll(ProfileRow::decode) { RestQuery.updateDisplayName(it.userId, validName) }
        return rows.firstOrNull()?.profile ?: throw AppError.Forbidden
    }
}

/** [PushService] on the `enable_push` / `disable_push` RPCs and `push_subscriptions` (docs/CONTRACTS.md §4, §7). */
internal class SupabasePushService(private val context: SupabaseContext) : PushService {
    override suspend fun currentTopic(): String? {
        val rows = context.rest.fetchAll(PushTopicRow::decode) { RestQuery.pushTopic(it.userId) }
        return rows.firstOrNull()?.topic
    }

    override suspend fun enable(): String =
        context.rest.fetch({ it.stringValue() ?: throw MalformedAnswer("expected a string") }) { RestQuery.rpc("enable_push") }

    override suspend fun disable() {
        context.rest.send { RestQuery.rpc("disable_push") }
    }
}
